Bi 168 add a way to get to jira tickets from GitHub

tungledao opened the pull request.

tungledao committed:
BI-168-Add-a-way-to-get-to-Jira-tickets-from-github

tungledao committed:
Merge from master

tungledao committed:
Comment will show only bug ID

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,58 @@
+			LOG.debug("Pull request title " + branch_name  + " does not contain any ticket of Cozmo project");
+		}
+		while (cozmo_matcher.find()){
+			String coz_id = cozmo_matcher.group().replaceAll("coz", jira_link + "coz-");
```
why is `coz-` here? I think it is `cozmo-`

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,58 @@
+		
+		Pattern cozmo_pattern = Pattern.compile(cozmo);
+		Matcher cozmo_matcher = cozmo_pattern.matcher(name);
+		if (!od_matcher.matches()){
```
need a space between `)` and `{`

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,58 @@
+		if (!od_matcher.matches()){
+			LOG.debug("Pull request title " + branch_name  + " does not contain any ticket of Cozmo project");
+		}
+		while (cozmo_matcher.find()){
```
need a space between `)` and `{`

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/ChangeStatusUpdater.java:
```diff
@@ -209,9 +209,30 @@ private String getComment(@NotNull RepositoryVersion version,
+            final String vcsBranch = version.getVcsBranch();
+            String title = "";
+            try {
+                title = api.getPullRequestTitle(repositoryOwner, repositoryName, vcsBranch);
```
should follow 2 space instead of 4

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,58 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+public class GetJiraTickets {
+	private static final Logger LOG = Logger.getInstance(GetJiraTickets.class.getName());
```
should follow 2 space instead of 4

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,58 @@
+		
+		Pattern bi_pattern = Pattern.compile(bi);
+		Matcher bi_matcher = bi_pattern.matcher(name);
+		if (!od_matcher.matches()){
```
need a space between `)` and `{`

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,58 @@
+		if (!od_matcher.matches()){
+			LOG.debug("Pull request title " + branch_name  + " does not contain any ticket of BI project");
+		}
+		while (bi_matcher.find()){
```
need a space between `)` and `{`

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,58 @@
+			String bi_id = bi_matcher.group().replaceAll("bi", jira_link + "bi-");
+			list_jira_link.add(bi_id);
+		}
+		if (!(list_jira_link.size()>0)){
```
need a space between `)` and `{`

manhtran1609 commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/api/GitHubApi.java:
```diff
@@ -67,6 +67,19 @@ String findPullRequestCommit(@NotNull String repoOwner,
+   */
+  @Nullable
+  String getPullRequestTitle(@NotNull String repoOwner,
+                              @NotNull String repoName,
```
please line up the @NotNull vertically so it's easier to read

tungledao committed:
Update BI-168:
- Merge master into branch
- Pass JIRA_LINK_CONF_NAME system parameter into script. Please add a new system parameter JIRA_LINK_CONF_NAME with value "https://ankiinc.atlassian.net/browse/" if using script GetJiraTickets.java
- The script support for all of patterns of ticket id (od 123, od-123, od123)
- The script works on the branch name

tungledao committed:
Update variables

tungledao committed:
Update GetJiraTickets.java

manhtran1609 commented:
@lip please help me to review

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,50 @@
+
+public class GetJiraTickets {
+  private static final Logger LOG = Logger.getInstance(GetJiraTickets.class.getName());
+  private static final String od = "od\\d*";
```
We should use upper case for constant

sonnguyen-logigear replied:
this pattern will match all candidate that contain **od**, we need a pattern to match **od d+** or **od-d+**

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/ChangeStatusUpdater.java:
```diff
@@ -209,6 +213,29 @@ private String getComment(@NotNull RepositoryVersion version,
+                comment.append(ticketLink);
+                comment.append(")\n");
+              }
+            }
```
We should make this as a function.

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,50 @@
+    while (odMatcher.find()) {
+      String odId = odMatcher.group().replaceAll("od", jiraLink + "od-");
+      listJiraLink.add(odId);
+    }
```
We should make it as a function, then reuse for cozmo, bi

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,50 @@
+      String biId = biMatcher.group().replaceAll("bi", jiraLink + "bi-");
+      listJiraLink.add(biId);
+    }
+    if (!(listJiraLink.size()>0)) {
```
just check size == 0

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/api/impl/GitHubApiImpl.java:
```diff
@@ -164,6 +164,29 @@ public String findPullRequestCommit(@NotNull String repoOwner,
+                                     @NotNull String repoName,
+                                     @NotNull String branchName) throws IOException {
+    final String pullRequestId = getPullRequestId(repoName, branchName);
+    if (pullRequestId == null) return null;
```
use if (condition) {
         //code
       } as standard

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/api/impl/GitHubApiImpl.java:
```diff
@@ -164,6 +164,29 @@ public String findPullRequestCommit(@NotNull String repoOwner,
+
+    final RepoInfo head = pullRequestInfo.head;
+    if (head != null) {
+      return head.ref;
```
We should declare a variable to hold return value, and using **return** at the end of method only

tungledao committed:
Update BI-168

tungledao committed:
Update to support for many id tickets

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/ChangeStatusUpdater.java:
```diff
@@ -191,6 +191,21 @@ private String getFriendlyDuration(final long seconds) {
+            for (int i = 0; i < listTicketLink.size(); i++) {
+              String ticketLink = listTicketLink.get(i);
+              String ticketId = ticketLink.substring(ticketLink.lastIndexOf("/") + 1).toUpperCase();
+              String commentLine = "[" + ticketId + "](" + ticketLink + ")\n";
```
[test](https://google.com)

sonnguyen-logigear replied:
I think it should a space instead of "\n" as : [OD-1](https://OD1) [OD-2](https://OD2)

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,37 @@
+  public static List<String> getListTicketLink(@NotNull String branchName, 
+                                               @NotNull String jiraLink) {
+
+    String name = branchName.toLowerCase().replaceAll("[^\\da-z]", "");
```
We only support patterns : od-123 and od 123, no need to replaceAll here

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/GetJiraTickets.java:
```diff
@@ -0,0 +1,37 @@
+    }
+    if (listTicketLink.size() == 0) {
+      LOG.debug("Pull request title " + branchName + " does not contain any ticket id");
+      listTicketLink = null;
```
We no need to assign listTicketLink to null, we can check by get size of it.

tungledao committed:
Update Pattern

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/ChangeStatusUpdater.java:
```diff
@@ -191,6 +191,21 @@ private String getFriendlyDuration(final long seconds) {
+
+          private StringBuilder getJiraTicketComment(@NotNull List<String> listTicketLink) {
+            StringBuilder jiraLinksComment = new StringBuilder();
+            for (int i = 0; i < listTicketLink.size(); i++) {
```
using for (String ticketLink :  listTicketLink) as our standard

sonnguyen-logigear commented on commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/github/ChangeStatusUpdater.java:
```diff
@@ -209,6 +224,22 @@ private String getComment(@NotNull RepositoryVersion version,
+            final String jiraLink = getJiraLink(build);
+            final List<String> listTicketLink = GetJiraTickets.getListTicketLink(refBranch, jiraLink);
+            if (listTicketLink.size() != 0) {
+              StringBuilder ticketLinkComment = getJiraTicketComment(listTicketLink);
```
We just need to you String instead of StringBuilder

tungledao committed:
Update to foreach
